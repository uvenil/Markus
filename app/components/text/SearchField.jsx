'use strict';

import React from 'react';
import muiThemeable from 'material-ui/styles/muiThemeable';
import Unique from '../../utils/Unique';
import Constants from '../../utils/Constants';
import { showTextBoxContextMenu } from '../../utils/ContextMenuUtils';
import assign from 'lodash.assign';
import isEmpty from 'lodash.isempty';

class SearchField extends React.Component {
    constructor(props) {
        super(props);

        this.textBoxId     = Unique.nextString('SearchField_textBoxId');
        this.clearButtonId = Unique.nextString('SearchField_clearButtonId');

        this._handleChange = value => {
            this.refs[this.clearButtonId].style.display = isEmpty(value) ? 'none' : 'block';

            if (this.props.onChange) {
                this.props.onChange(value);
            }
        };

        this._handleClear = () => {
            const textBox = this.refs[this.textBoxId];

            textBox.value = '';
            textBox.focus();

            this._handleChange('');
        };
    }

    render() {
        const style = { width : '100%', position : 'relative', pointerEvents : this.props.disabled ? 'none' : 'auto' };

        return (
            <div
                className={this.props.className}
                style={assign(style, this.props.style)}>
                <i
                    className="fa fa-fw fa-search"
                    style={{ position : 'absolute', left : Constants.PADDING_X0 + Constants.PADDING_X1, top : Constants.PADDING_X0 + Constants.PADDING_X1, fontSize : 14, color : this.props.muiTheme.palette.disabledColor, cursor : 'default' }} />
                <input
                    ref={this.textBoxId}
                    type="text"
                    placeholder={this.props.hintText}
                    style={{ width : 'calc(100% - ' + (Constants.PADDING_X1 * 2 + Constants.PADDING_X2 * 2) + 'px)', paddingLeft : Constants.PADDING_X1 + Constants.PADDING_X2, paddingRight : Constants.PADDING_X1 + Constants.PADDING_X2, paddingTop : Constants.PADDING_X0, paddingBottom : Constants.PADDING_X0, WebkitUserSelect : 'none', color : this.props.muiTheme.palette.textColor, border : 'none', outline : 'none' }}
                    onChange={event => this._handleChange(event.target.value)}
                    onMouseDown={event => {
                        if (event.nativeEvent.button === 2) {
                            showTextBoxContextMenu();
                        }
                    }} />
                <i
                    ref={this.clearButtonId}
                    className="fa fa-fw fa-times-circle"
                    style={{ display : 'none', position : 'absolute', right : Constants.PADDING_X0 + Constants.PADDING_X1, top : Constants.PADDING_X0 + Constants.PADDING_X1, fontSize : 14, color : this.props.muiTheme.palette.disabledColor, cursor : 'pointer' }}
                    onClick={() => this._handleClear()} />
            </div>
        );
    }
}

SearchField.propTypes = {
    hintText  : React.PropTypes.string,
    className : React.PropTypes.string,
    style     : React.PropTypes.object,
    disabled  : React.PropTypes.bool,
    onChange  : React.PropTypes.func
};

SearchField.defaultProps = {
    hintText : 'Search',
    disabled : false
};

export default muiThemeable()(SearchField);
