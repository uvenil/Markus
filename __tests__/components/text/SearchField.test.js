'use strict';

import React from 'react';
import Renderer from 'react-test-renderer'
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import SearchField from '../../../app/components/text/SearchField.jsx';
import { mount } from 'enzyme';

jest.mock('../../../app/utils/ContextMenuUtils');

it('renders correctly', () => {
    const component = Renderer.create(
        <MuiThemeProvider muiTheme={getMuiTheme(lightBaseTheme)}>
            <SearchField hintText="Search" />
        </MuiThemeProvider>
    ).toJSON();

    expect(component).toMatchSnapshot();
});

it('triggers onChange after a key press, and clear text after clicking clear button', () => {
    let searchFieldValue = '';

    const component = mount(
        <MuiThemeProvider muiTheme={getMuiTheme(lightBaseTheme)}>
            <SearchField
                hintText="Search"
                onChange={value => searchFieldValue = value} />
        </MuiThemeProvider>
    );

    const input = component.find('#SearchField_textBoxId');

    input.simulate('focus');
    input.simulate('change', {
        target : {
            value : 'a'
        }
    });

    expect(searchFieldValue).toEqual('a');

    const clearButton = component.find('#SearchField_clearButtonId');
    clearButton.simulate('click');

    expect(searchFieldValue).toEqual('');
});
