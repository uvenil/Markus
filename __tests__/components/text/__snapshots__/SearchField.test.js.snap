// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className={undefined}
  style={
    Object {
      "pointerEvents": "auto",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <i
    className="fa fa-fw fa-search"
    style={
      Object {
        "color": "rgba(0, 0, 0, 0.3)",
        "cursor": "default",
        "fontSize": 14,
        "left": 12,
        "position": "absolute",
        "top": 12,
      }
    }
  />
  <input
    id="SearchField_textBoxId"
    onChange={[Function]}
    onMouseDown={[Function]}
    placeholder="Search"
    style={
      Object {
        "WebkitUserSelect": "none",
        "border": "none",
        "color": "rgba(0, 0, 0, 0.87)",
        "outline": "none",
        "paddingBottom": 4,
        "paddingLeft": 24,
        "paddingRight": 24,
        "paddingTop": 4,
        "width": "calc(100% - 48px)",
      }
    }
    type="text"
  />
  <i
    className="fa fa-fw fa-times-circle"
    id="SearchField_clearButtonId"
    onClick={[Function]}
    style={
      Object {
        "color": "rgba(0, 0, 0, 0.3)",
        "cursor": "pointer",
        "display": "none",
        "fontSize": 14,
        "position": "absolute",
        "right": 12,
        "top": 12,
      }
    }
  />
</div>
`;
